### ISSUE TITLE
linter: `no-unnecessary-condition` false positive on string equality check with generic type

### ISSUE THREAD

**Issue by DreierF** [oxc-project/oxc#20487]
### What version of Oxlint are you using?

1.56.0 / 0.17.0

### What command did you run?

oxlint --type-aware

### What does your `.oxlintrc.json` (or `oxlint.config.ts`) config file look like?

```json
{
  "plugins": ["typescript"],
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

### What happened?

`no-unnecessary-condition` reports "This condition will always return the same value since the types have no overlap" on `value !== ''` where `value` has type `NonNullable<T>` (after a `!= null` guard on `T | null | undefined`).

Oxlint considers `NonNullable<T>` and `""` to have no overlap, but when `T` is instantiated as `string`, `NonNullable<T>` is just `string`, which can absolutely be `""`.

`typescript-eslint` does not report this case.

### Minimal reproduction

`repro.ts`:
```typescript
function getValueOrDefault<T>(value: T | null | undefined, defaultValue: T): T {
	if (value != null && value !== '') {
		return value;
	}
	return defaultValue;
}

const _result = getValueOrDefault<string>('', 'fallback');
```

`tsconfig.json`:
```json
{
  "compilerOptions": {
    "strict": true,
    "target": "ES2020",
    "module": "ES2020",
    "moduleResolution": "bundler"
  },
  "include": ["*.ts"]
}
```

### Expected behavior

No error. `T` is unconstrained and can be `string`, making `value !== ''` a meaningful check.

### Actual behavior

```
  x typescript-eslint(no-unnecessary-condition): This condition will always return the same value since the types have no overlap.
   ,-[repro.ts:2:22]
 1 | function getValueOrDefault<T>(value: T | null | undefined, defaultValue: T): T {
 2 |     if (value != null && value !== '') {
   :                         ^^^|^^    ^|^
   :                            |       `-- Type: ""
   :                            `-- Type: NonNullable<T>
 3 |         return value;
   `----
```

The root cause is that oxlint treats `NonNullable<T>` (where `T` is unconstrained) as incompatible with the string literal `""`, when `T` can be `string` itself.
